### PR TITLE
Add CSVFileWrapper

### PR DESCRIPTION
**Pull request by OlgaOvcharenko**
This PR adds CSV file wrapper and corresponding tests, and addresses issue #235. 
CSVs are read as pandas DataFrames in methods as `get_sample` etc. Therefore, pandas was added. 
All data is returned as bytes or list of bytes as discussed.

### CODE REVIEW

File: modyn/storage/internal/file_wrapper/csv_file_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            self.extension = ".csv"
+
+        self.encoding = file_wrapper_config["encoding"]
+        if self.encoding is None or self.encoding == "":
+            self.extension = "utf-8"
+
+        self.header = file_wrapper_config["header"]
+        if self.header is None:
+            self.header = None
+
```
[the remark this answers is not in the document]
Comment (OlgaOvcharenko) [reply modyn/storage/internal/file_wrapper/csv_file_wrapper.py:55]: Fixed.

File: modyn/storage/internal/file_wrapper/csv_file_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    def get_number_of_samples(self) -> int:
+        """Get number of samples in file.
+
+        Returns:
+            int: Number of samples in file
+        """
+        i = 0
+        with open(self.file_path, "r", encoding=self.encoding) as file_pointer:
+            for _ in file_pointer:
+                i += 1
```
[the remark this answers is not in the document]
Comment (OlgaOvcharenko) [reply modyn/storage/internal/file_wrapper/csv_file_wrapper.py:76]: Okay, but in this particular case, the list comprehension is not more efficient for multiple reasons: allocating the list, then summing.
Here is the average time (in sec) for this particular part of the code for 50 runs for 4 million rows: 
List comprehension + sum over the list - 1.0179
Your version - 1.3749
**Loop from PR - 0.8268**

If you still prefer list comprehension, I can put it in, but I do not think it is more efficient.

```
n_runs = 50
runs = []
for method in range(1, 4):
    for i in range(n_runs):
        if method == 1:
            t1 = time.time()
            with open(self.file_path, "r", encoding=self.encoding) as file:
                num_lines = sum([1 for line in file if line.rstrip()])
            runs.append(time.time() - t1)
        elif method == 2:
            t1 = time.time()
            num_lines = sum(1 for line in open(self.file_path, "r", encoding=self.encoding) if line.rstrip())
            runs.append(time.time() - t1)
        else:
            t1 = time.time()
            num_lines = 0
            with open(self.file_path, "r", encoding=self.encoding) as file:
                for _ in file:
                    num_lines += 1
            runs.append(time.time() - t1)
    runs = runs[2:]
    print(method)
    print(sum(runs)/len(runs))
    runs = []
```

File: modyn/storage/internal/file_wrapper/csv_file_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            index (int): Index
+
+        Raises:
+            IndexError: If the index is out of bounds
+
+        Returns:
+            object: Label for the sample
+        """
+        data_frame = pd.read_csv(self.file_path, sep=self.delim, header=self.header, encoding=self.encoding)
+        return bytes(data_frame.columns[index], self.encoding)
```
[the remark this answers is not in the document]
Comment (OlgaOvcharenko) [reply modyn/storage/internal/file_wrapper/csv_file_wrapper.py:92]: Tried to fix it, but would like to discuss it.

File: modyn/storage/internal/file_wrapper/csv_file_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+    def get_all_labels(self) -> list[bytes]:
+        """Returns a list of all labels of all samples in the file.
+
+        Returns:
+            list: List of labels
+        """
+        data_frame = pd.read_csv(self.file_path, sep=self.delim, header=self.header, encoding=self.encoding)
+        return [bytes(col, self.encoding) for col in data_frame.columns.tolist()]
```
[the remark this answers is not in the document]
Comment (OlgaOvcharenko) [reply modyn/storage/internal/file_wrapper/csv_file_wrapper.py:101]: Fixed as above.

File: modyn/storage/internal/file_wrapper/csv_file_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+        Raises:
+            IndexError: If the index is out of bounds
+
+        Returns:
+            bytes: Sample
+        """
+        data_frame = pd.read_csv(self.file_path, sep=self.delim, header=self.header, encoding=self.encoding)
+        samples = data_frame.iloc[indices].values.tolist()
+        res = [[bytes(val, self.encoding) for val in row] for row in samples]
```
[the remark this answers is not in the document]
Comment (OlgaOvcharenko) [reply modyn/storage/internal/file_wrapper/csv_file_wrapper.py:149]: This leads to just a string that contains multiple values, is it what you intend?
For example, b"hero,description\nbatman,uses technology\n".

File: modyn/storage/internal/file_wrapper/csv_file_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            features_indices (list): List of feature names (str) and indices (int) of the required samples
+
+        Raises:
+            IndexError: If the index is out of bounds
+
+        Returns:
+            bytes: Sample
+        """
+        data_frame = pd.read_csv(self.file_path, sep=self.delim, header=self.header, encoding=self.encoding)
+        samples = [bytes(data_frame.get_value(index, feature)) for feature, index in features_indices]
```
[the remark this answers is not in the document]
Comment (OlgaOvcharenko) [reply modyn/storage/internal/file_wrapper/csv_file_wrapper.py:166]: Removed it.

File: modyn/storage/internal/file_wrapper/csv_file_wrapper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/storage/internal/file_wrapper/csv_file_wrapper.py:68]: Ohh thank you so much for benchmarking this. Then let's keep it the way you suggested. Thanks for the thorough investigation!

**Comment by MaxiBoether** [comment]
Closing for now, as @francescodeaglio will take over the CSV Wrapper for now. We can always open a new PR later :)

**Pull request closed without merging**